lib/dustoff/accounts.ex

```diff
@@ -0,0 +1,229 @@
+defmodule Dustoff.Accounts do
+  @moduledoc """
+  Provides functions for managing user accounts, authentication and sessions.
+  """
+
+  alias Dustoff.Accounts.User
+  alias Dustoff.Accounts.UserNotifier
+  alias Dustoff.Accounts.UserToken
+  alias Dustoff.Repo
+
+  @doc """
+  Gets a `Dustoff.Accounts.User` entity by email.
+  """
+  @spec get_user_by_email(email :: String.t()) :: User.t() | nil
+  def get_user_by_email(email) when is_binary(email) do
+    Repo.get_by(User, email: email)
+  end
+
+  @doc """
+  Gets a `Dustoff.Accounts.User` entity by email and password.
+  """
+  @spec get_user_by_email_and_password(
+          email :: String.t(),
+          password :: String.t()
+        ) :: User.t() | nil
+  def get_user_by_email_and_password(email, password)
+      when is_binary(email) and is_binary(password) do
+    user = Repo.get_by(User, email: email)
+    if User.valid_password?(user, password), do: user
+  end
+
+  @doc """
+  Gets a single `Dustoff.Accounts.User` entity.
+
+  Raises `Ecto.NoResultsError` if the User does not exist.
+  """
+  @spec get_user!(User.id()) :: User.t()
+  def get_user!(id), do: Repo.get!(User, id)
+
+  ## User registration
+
+  @doc """
+  Registers a user.
+  """
+  @spec register_user(attrs :: map()) :: {:ok, User.t()} | {:error, Ecto.Changeset.t()}
```

Comment:
This return type is with intent not `User.changeset()` because that type is meant for repo-backed entities.